client/protocols/guidance.py

```diff
@@ -0,0 +1,178 @@
+OK_GUIDANCE = "tests/.ok_guidance"
+
+lambda_string_key_to_func = {
+    'none': lambda info, strMisU: None, 
+    'ki': lambda info, strMisU: info['ki'],
+    'reteach': lambda info, strMisU: info['dictMisU2ReteachMsg'].get(strMisU)
+}
+  
+# If student forces guidance messages to show, we will assign treatment group number below
+guidance_flag_dict = [1]
+
+class GuidanceProtocol(models.Protocol):
+    def __init__(self,working_dir,debug = -1):
+        self.tg_id = -1
+        self.msg_id = -1
+        self.misU_count = {}
+        self.working_dir = working_dir
+        self.debug = debug
+        try:
+            with open(working_dir + OK_GUIDANCE,"r") as f:
+                self.guidance_json = json.load(f)
+            self.error = False
+        except:
+            self.error = True
+
+    def guidance_msg(self,unique_id, input_lines,access_token,hash_key,guidance_flag):
+        if self.error:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        shorten_unique_id = unique_id[unique_id.index('\n')+1:]
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        dict_info = wa_2_dict_info.get(repr(input_lines))
+
+        # If this wrong answer is not in the JSON file, display default message
+        if not dict_info:            
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+        if(self.debug != -1):
+            self.tg_id = self.debug
+        else:
+            self.get_tg(access_token,guidance_flag)
+
+        if self.tg_id == -1: 
+            # If self.tg_id == -1, some errors happen when trying to access server
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_string_key = self.guidance_json['dictTg2Func'].get(str(self.tg_id))
+
+        if not lambda_string_key:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
+
+        lst_mis_u = dict_info.get('lstMisU')
+
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        self.misU_count = self.update_misUcounts(hash_key, lst_mis_u, repr(input_lines),shorten_unique_id)
+
+        thresold = self.guidance_json['wrongAnsThresh']
+
+        """
+        Update count file using the helper. Each misunderstanding igets one count.
+
+        Get the updated acculamted counts as a dictionary: key: mis_u, value: counts.
+        
+        Called that dictionary self.misU_count
+        """
```

Comment:
Yeah, it was the stub code of Spenser's part when I was developing on my branch. Should we just get rid of this completely? @ksteph